# Import Libraries

In [3]:
import pandas as pd
import numpy as np
import os

# Importing Data

In [6]:
path = r'C:\Users\shaiv\Downloads\Instacart Basket Analysis'

In [8]:
path

'C:\\Users\\shaiv\\Downloads\\Instacart Basket Analysis'

In [13]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [15]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [27]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

# Wrangling Data

### 02 Find another identifier variable in the df_ords dataframe 

##### Orders Data Frame

In [33]:
# dropping eval_set column from orders data frame
df_ords = df_ords.drop(columns = ['eval_set'])

In [35]:
#changing data type for order and user ID in orders data frame
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')

##### Departments Data Frame

In [38]:
# transposing df_dep
df_dep_t = df_dep.T

##### Change Header of Departments df

In [41]:
# Take 1st row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [43]:
# Copying df_dep_t from row 1 onward
df_dep_t_new = df_dep_t[1:]

In [45]:
# Set the header row as the df header
df_dep_t_new.columns = new_header

In [47]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

###### Creating Data Dictionary

In [50]:
data_dict = df_dep_t_new.to_dict('index')

##### Changing variable to suitable format in df_ords

In [67]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   order_number            int64  
 3   order_dow               int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(2)
memory usage: 156.6+ MB


In [75]:
# Changing order_number to string as it is not required for stats
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [77]:
df_ords.describe()

order_dow  order_hour_of_day  days_since_prior_order
count  3.421083e+06       3.421083e+06            3.214874e+06
mean   2.776219e+00       1.345202e+01            1.111484e+01
std    2.046829e+00       4.226088e+00            9.206737e+00
min    0.000000e+00       0.000000e+00            0.000000e+00
25%    1.000000e+00       1.000000e+01            4.000000e+00
50%    3.000000e+00       1.300000e+01            7.000000e+00
75%    5.000000e+00       1.600000e+01            1.500000e+01
max    6.000000e+00       2.300000e+01            3.000000e+01

### 03 Renaming column day of the week in orders data frame

In [81]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_the_week'}, inplace = False)

order_id user_id order_number  orders_day_of_the_week  \
0        2539329       1            1                       2   
1        2398795       1            2                       3   
2         473747       1            3                       3   
3        2254736       1            4                       4   
4         431534       1            5                       4   
...          ...     ...          ...                     ...   
3421078  2266710  206209           10                       5   
3421079  1854736  206209           11                       4   
3421080   626363  206209           12                       1   
3421081  2977660  206209           13                       1   
3421082   272231  206209           14                       6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

### 04 Busiest hour of the day

In [84]:
df_ords['order_hour_of_day'].value_counts(dropna = True)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

##### The busiest hour of the day is 10am.

### 05 What "4" means in the data dictionary

In [88]:
print(data_dict.get('4'))

{'department': 'produce'}


##### The department_id "4" corresponds to the category "produce".

### 06 Creating subset for breakfast items

##### From data dictionary department_id for "breakfast" is 14.

In [93]:
# Creating subset for breakfast
df_prods_breakfast = df_prods.loc[df_prods['department_id'] == 14]

In [95]:
df_prods_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

### 07 Creating subset for dinner party items

In [102]:
# Creating subset for alcohol, deli, beverages and meat/seafood
df_prods_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [106]:
df_prods_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### 08 Number of rows in dinner party subset

In [109]:
df_prods_dinner_party.shape

(7650, 5)

##### There are 7650 rows in dinner party subset

### 09 Information on user_id 1

In [115]:
df_user_id_1 = df_ords.loc[df_ords['user_id'] == '1']
df_user_id_1

order_id user_id order_number  order_dow  order_hour_of_day  \
0   2539329       1            1          2                  8   
1   2398795       1            2          3                  7   
2    473747       1            3          3                 12   
3   2254736       1            4          4                  7   
4    431534       1            5          4                 15   
5   3367565       1            6          2                  7   
6    550135       1            7          1                  9   
7   3108588       1            8          1                 14   
8   2295261       1            9          1                 16   
9   2550362       1           10          4                  8   
10  1187899       1           11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### 10 Stats on user_id 1 behaviour

In [118]:
df_user_id_1.describe()

order_dow  order_hour_of_day  days_since_prior_order
count  11.000000          11.000000               10.000000
mean    2.636364          10.090909               19.000000
std     1.286291           3.477198                9.030811
min     1.000000           7.000000                0.000000
25%     1.500000           7.500000               14.250000
50%     3.000000           8.000000               19.500000
75%     4.000000          13.000000               26.250000
max     4.000000          16.000000               30.000000

##### This user exhibits a consistent ordering pattern, placing orders exclusively between Sunday and Wednesday, with no activity from Thursday to Saturday. A significant portion of their orders (over 50%) are placed at 8:00 AM or earlier, showcasing a preference for early transactions. On average, they place an order every 19 days, with a maximum gap of 30 days between orders. Notably, the user has demonstrated the ability to place multiple orders in a single day, having done so at least twice.

### 11 Exporting order data frame

In [124]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

### 12 Exporting departments data frame

In [128]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))